### PR TITLE
Closed the perf gap of resnet and enabled refit

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py [hunk 1]
```diff
@@ -48,89 +49,160 @@ def batch_norm(
...
+            isinstance(running_var, torch.Tensor),
+        ]
+    ):
+        if weight is None:
+            weight = 1.0
+
+        if bias is None:
+            bias = 0.0
+
+        if running_mean is None:
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: add a docstring in this function explaining the batchnorm folding and why it is the default behavior for performance reasons for future reference.

**Review by peri044 — COMMENTED** [review]
minor comment. LGTM

File: py/torch_tensorrt/dynamo/_refit.py [hunk 2]
```diff
@@ -90,7 +91,18 @@ def construct_refit_mapping_from_weight_name_map(
 ) -> dict[Any, Any]:
     engine_weight_map = {}
     for engine_weight_name, (sd_weight_name, np_weight_type) in weight_name_map.items():
-        if sd_weight_name not in state_dict:
+        if engine_weight_name.split(" ")[-1] in ["SCALE", "SHIFT"]:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:98]: We should abstract this imo. Like if there are any weight types that require constant folding in converter this should be associated with the converter. Then the refit system will just iterate through all these constant fold operations. Ideally the converter can use the same implementation

Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_refit.py:98]: Currently BN is the only one. Do you think we should have a constant_fold function and have `refit` and `conversion` call that function?

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py [hunk 3]
```diff
@@ -47,90 +48,163 @@ def batch_norm(
...
+    # In this way, the batch norm layer will be fused with the Convolution layer and get a performance boost.
+    if all(
+        [
+            isinstance(weight, torch.Tensor),
+            isinstance(bias, torch.Tensor),
+            isinstance(running_mean, torch.Tensor),
+            isinstance(running_var, torch.Tensor),
+        ]
+    ):
+        if weight is None:
```
Comment (HolyWu) [inline py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: I think this check is redundant. If `weight` was None, then the `isinstance(weight, torch.Tensor)` above would already return `False`, and hence it wouldn't enter this branch at all.

File: py/torch_tensorrt/dynamo/_refit.py (hunk 2 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_refit.py:98]: yeah we should think of some sort of abstraction where the constant folding process is defined once for both refit and conversion

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py (hunk 3 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: Think the correct check on L54 is more like if all parameters are not ITensors. If one is none and the rest are torch tensors or floats we can still do constant folding. but if one of the parameters is a ITensor then we cant

Comment (cehongwang) [reply py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: IDK when I don't have the check the pre-commit gives me errors

File: py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py [hunk 4]
```diff
@@ -47,90 +48,164 @@ def batch_norm(
 
     # Save the original output shape for later use
     output_shape = input.shape
+    # We perform constant folding for batch norm when the weight, bias, running_mean, and running_var are all tensors.
+    # Batch norm operation can be fused into a single layer, which is more efficient than the original implementation.
+    # In this way, the batch norm layer will be fused with the Convolution layer and get a performance boost.
+    if not all(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: @cehongwang, weight, bias, etc can be none but we can still do constant folding. We just need to check that they are not ITensors: https://github.com/pytorch/pytorch/blob/f16053f0c9a09fa337fbf85aaf64f88712b8dcdb/aten/src/ATen/native/native_functions.yaml#L1093

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/impl/normalization/ops.py]: So this check should probably be `if any([isinstance(weight, trt.ITensor), ...])`

**Review by narendasan — APPROVED** [review]
LGTM
